util/pyclient/bft_client.py

```diff
@@ -284,7 +284,7 @@ def _process_received_msg(self, data, sender, replicas_addr, required_replies, c
             self.replies_manager.add_reply(rsi_msg)
             if self.replies_manager.has_quorum_on_all(required_replies):
                 self.replies = self.replies_manager.get_all_replies()
-                self.rsi_replies = self.replies_manager.get_rsi_replies(rsi_msg.get_matched_reply_key())
```

Comment:
Could you please explain the Apollo RSI changes?

Reply:
Currently, the common data contain also the primary. However, we want to enable read requests to read only replicas, and ro replica doesn't know who the primary is

Reply:
Does the primary now become part of RSI?

Reply:
I'm not sure this change is correct. The knowledge of the primary is required for normal operation of the client. The client should only send to the primary for write operations, unless there is a timeout. Sending to all replicas should only be done on timeout to ensure a primary is not malicious, otherwise we are sending more traffic than necessary. It shouldn't induce view change, but it's still different from the behavior of the real C++ client, and doesn't seem right to me. 

Excluding the primary is probably fine for read-only operations, but not write operations.

Reply:
@andrewjstone, the replicas still send the write requests to the primary only. This is just an implementation change. We collect the results from the replicas for a specific sequence number under a common key (as the reply may contain RSI). This change means that the primary number is not part of this common data, but it is still part of the reply itself, once the client decides the replies can be delivered up, it delivers also the primary ID.
We removed it from the common data, because the RORs don't have the knowledge on who the primary is, and we would like to collect their responses together with the committers reply.

Reply:
@andrewjstone in any case, I revreted this change. I will perform vast changes once I will merge the Apollo tests for this